tsl/src/chunk.c

```diff
@@ -1180,3 +1186,579 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+typedef struct RelationSplitInfo
+{
+	BulkInsertState bistate;
+	TupleTableSlot *dstslot;
+	RewriteState rwstate;
+	Relation rel;
+	Datum *values;
+	bool *isnull;
+	/*
+	 * Tuple mapping is needed in case the old relation has dropped
+	 * columns. New relations (as result of split) are "clean" without dropped
+	 * columns. The tuple map converts tuples between the source and
+	 * destination chunks.
+	 */
+	TupleConversionMap *tupmap;
+} RelationSplitInfo;
+
+static RelationSplitInfo *
+relation_split_info_create(Relation splitrel, Relation rel, struct VacuumCutoffs *cutoffs)
+{
+	RelationSplitInfo *rsi = palloc0(sizeof(RelationSplitInfo));
+
+	rsi->rel = rel;
+	rsi->bistate = GetBulkInsertState();
+	rsi->rwstate = begin_heap_rewrite(splitrel,
+									  rel,
+									  cutoffs->OldestXmin,
+									  cutoffs->FreezeLimit,
+									  cutoffs->MultiXactCutoff);
+
+	rsi->tupmap = convert_tuples_by_name(RelationGetDescr(splitrel), RelationGetDescr(rel));
+
+	/* Create tuple slot for new partition. */
+	rsi->dstslot = MakeSingleTupleTableSlot(RelationGetDescr(rel), table_slot_callbacks(rel));
+	ExecStoreAllNullTuple(rsi->dstslot);
+
+	rsi->values = (Datum *) palloc0(RelationGetDescr(splitrel)->natts * sizeof(Datum));
+	rsi->isnull = (bool *) palloc0(RelationGetDescr(splitrel)->natts * sizeof(bool));
+
+	return rsi;
+}
+
+static void
+relation_split_info_free(RelationSplitInfo *rsi, int ti_options)
+{
+	ExecDropSingleTupleTableSlot(rsi->dstslot);
+	FreeBulkInsertState(rsi->bistate);
+	table_finish_bulk_insert(rsi->rel, ti_options);
+	end_heap_rewrite(rsi->rwstate);
+	table_close(rsi->rel, NoLock);
+	pfree(rsi->values);
+	pfree(rsi->isnull);
+
+	if (rsi->tupmap)
+		free_conversion_map(rsi->tupmap);
+
+	rsi->rel = NULL;
+	rsi->bistate = NULL;
+	rsi->dstslot = NULL;
+	rsi->tupmap = NULL;
+	rsi->values = NULL;
+	rsi->isnull = NULL;
+	pfree(rsi);
+}
+
+/*
+ * Reconstruct and rewrite the given tuple.
+ *
+ * Mostly taken from heapam module.
+ *
+ * When splitting a relation in two, the old relation is retained for one of
+ * the result relations while the other is created new. This might lead to a
+ * situation where the two result relations have different attribute mappings
+ * because the old one could have dropped columns while the new one is "clean"
+ * without dropped columns. Therefore, the rewrite function needs to account
+ * for this when the tuple is rewritten.
+ */
+static void
+reform_and_rewrite_tuple(HeapTuple tuple, Relation splitrel, RelationSplitInfo *rsi)
+{
+	TupleDesc oldTupDesc = RelationGetDescr(splitrel);
+	TupleDesc newTupDesc = RelationGetDescr(rsi->rel);
```

Comment:
Let's call them something like "target relation" and "source relation" for clarity, instead of `splitrel` and `rel`.

Reply:
Updated.